### DISCUSSION TITLE
improve limiter by  ja3 fingerprint from HTTPS server

### DISCUSSION THREAD

**Discussion by return42** [searxng/searxng#2264] [Ideas]
To improve the [limiter](https://docs.searxng.org/src/searx.plugins.limiter.html#limiter-plugin), it would be cool to get a [JA3 fingerprint](https://engineering.salesforce.com/tls-fingerprinting-with-ja3-and-ja3s-247362855967/) from the [reverse proxy](https://docs.searxng.org/admin/architecture.html).

The modules I know are:

- https://github.com/phuslu/nginx-ssl-fingerprint
- https://docs.trafficserver.apache.org/en/9.1.x/admin-guide/plugins/ja3_fingerprint.en.html

In the limiter a `X-JA3-Sig` (or similar) field can be read from the `request.headers`  ..

https://github.com/searxng/searxng/blob/9e9f57e48bbc3c07d49dad57bb851f59f1c7def8/searx/plugins/limiter.py#L39-L43

and its value can be used to detect burst requests.  Currently the limiter uses the `X-Forwarded-For` value, which only works on the basis of an IP. With `X-JA3-Sig` one could more precisely track the requests of bots.

https://github.com/searxng/searxng/blob/9e9f57e48bbc3c07d49dad57bb851f59f1c7def8/searx/plugins/limiter.py#L51-L52

Does anyone already have experience with JA3 and the HTTPS servers?

**Comment by unixfox** [comment]
A project that globalize all the fingerprinting on TLS and HTTP2: https://pawprint.dev/
Source code: https://github.com/picoHz/pawprint

Do note that in my opinion, implementing any kind of fingerprinting is against keeping the privacy of the user.

**Reply by Solirs** [reply]
I agree that fingerprinting in any way is not the most ethical thing to implement, imo we should hold off on it at least as long as instances aren't heavily targeted by bots.

**Reply by return42** [reply]
> implementing any kind of fingerprinting is against keeping the privacy of the user.

really "any kind"? .. 

The fingerprint is a hash value .. we store this hash value at maximum for 600sec / in the current implementation of the limiter we store the hashed IP for a maximum of 600sec.

@unixfox do you think this would be critical?

**Reply by unixfox** [reply]
Usually fingerprinting is so much more personal and direct, a hash won't change it's tied to a specific device.
A public IP address can be shared by multiple devices and is not directly associated with a device. Even on IPv6, the device usually allocates a random IPv6 address that change every X times or when the device is rebooted.

Storing some info like the IP address about the users even when hashed already decrease the privacy, it's a fact keeping logs of the users even for a short period of time decrease the privacy because a trace of a user activity is left on a system. Using fingerprinting means that for a specific device, even if the device switches to another internet connection, the hash will always be the same.
So in a sense this could be interpreted as precise tracking, something that we are fighting against since SearX was created.

**BUT** like I wrote in https://github.com/searxng/searxng/discussions/2264#discussioncomment-5375307, usually TLS and HTTP2 fingerprinting target a specific version of a piece of software and **not a device**, so this may be an okay usage in this case.

So as long as we are using information that can be shared between multiples devices, this would be ok. Do note that this could be much more of a privacy issue if we merge two kinds of information together, as for example, there won't be that many people using Chrome version 166 on a specific IP address. The final "hash" would be less generic and much more precise to a specific user.

**Comment by unixfox** [comment]
The issue with this approach is that the hash given by TLS fingerprinting is specific to an HTTP client like curl at X version. Same goes for HTTP/2 fingerprints

So without a database referencing all the possibles hashes corresponding to an HTTP client that can be used for automation/bot, this method is useless. Let's say the hash for chrome is AAA, if you rate limit based only on this hash, you would rate limit pretty quickly all the users using that specific version of chrome.

We could maintain a database where we manually reference this kind of hashes but this could be very time consuming.

**Comment by unixfox** [comment]
Some insights about the cloudflare anti bot: https://nitter.net/BillDemirkapi/status/1663580689969577990

Maybe there are some things that could be interesting for our limiter like HTTP2 checking.
